app/grandchallenge/algorithms/views.py

```diff
@@ -335,6 +338,67 @@ def get_context_data(self, *args, **kwargs):
         return context
 
 
+NOT_IN_REGISTRY_ERROR = "NOT_IN_REGISTRY"
+
+
+class AlgorithmImageMarkDesired(
+    LoginRequiredMixin,
+    ObjectPermissionRequiredMixin,
+    SuccessMessageMixin,
+    FormView,
+):
+    permission_required = "algorithms.change_algorithm"
+    raise_exception = True
+    form_class = ImageMarkDesiredForm
+    success_message = "Image successfully updated."
+    template_name = "algorithms/mark_desired_version.html"
+
+    @cached_property
+    def algorithm(self):
+        return self.algorithm_image.algorithm
+
+    @cached_property
+    def algorithm_image(self):
+        return get_object_or_404(AlgorithmImage, pk=self.kwargs["pk"])
+
+    def get_permission_object(self):
+        return self.algorithm
+
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        kwargs.update({"image": self.algorithm_image})
+        return kwargs
+
+    def form_valid(self, form):
+        response = super().form_valid(form=form)
+        self.algorithm_image.mark_desired_version()
+        return response
+
+    def form_invalid(self, form):
+        super().form_invalid(form=form)
+        if form.has_error(NON_FIELD_ERRORS, code=NOT_IN_REGISTRY_ERROR):
+            self.algorithm_image.import_status = ImportStatusChoices.QUEUED
+            self.algorithm_image.save()
+            upload_to_registry_and_sagemaker.signature(
+                kwargs={
+                    "app_label": self.algorithm_image._meta.app_label,
+                    "model_name": self.algorithm_image._meta.model_name,
+                    "pk": self.algorithm_image.pk,
+                    "mark_as_desired": True,
+                }
+            ).apply_async()
+            messages.success(
+                self.request,
+                "Image validation and upload to registry in progress. It can take a while for this image to become active, please be patient.",
+            )
```

Comment:
You should find a way to handle this in `form_valid`. The form should only be concerned about if the image can be marked as desired, then importing it into the registry should happen in `form_valid`. Definitely don't raise a validation error on the form just to ignore it and do something with the form data anyway.

Reply:
But then you're splitting the logic over two places. We would then need to check `can_execute` and `image_upload_in_progress` both in the form and in the `form_valid` method.

Reply:
This is about responsibilities and separation of concerns. The Form is responsible for defining and validating what to change, the view is how to do it. The value of `can_execute` is cached (or should be if it is not), so it doesn't cost anything to call it twice.